src/rosdiscover/observer/ros1.py

```diff
@@ -0,0 +1,304 @@
+                action_server_candidates = dict()
+                action_client_candidates = dict()
+
+                update_node_contexts_with_topics(nodecontexts, info.publishers.items(), True,
+                                                 action_client_candidates,
+                                                 action_server_candidates, ros)
+                update_node_contexts_with_topics(nodecontexts, info.subscribers.items(), False,
+                                                 action_client_candidates,
+                                                 action_server_candidates, ros)
+
+                for service, nodes in info.services.items():
+                    if service.split('/')[-1] not in _SERVICES_TO_FILTER_OUT:
+                        for node in nodes:
+                            nodecontexts[node].provide(service,
+                                                       ros.services[service].format.fullname)
+
+                # Check if action candidates are complete (i.e., have all their topics)
+                # and add action if they are, or add the topics back in if they're not
+                for node, action_servers in action_server_candidates.items():
+                    for action_server in action_servers.values():
+                        if action_server.is_complete():
+                            nodecontexts[node].action_server(action_server.name, action_server.fmt)
+                        else:
+                            logger.warning(f"Action {action_server._action_name} is incomplete")
+                            logger.debug(
+                                f"goal({action_server.goal}), cancel({action_server.cancel}), "
+                                f"status({action_server.status}), feedback("
+                                f"{action_server.feedback}), result({action_server.result}")
+                            action_server.add_unfinished_to_context(nodecontexts[node])
+
+                for node, action_clients in action_client_candidates.items():
+                    for action_client in action_clients.values():
+                        if action_client.is_complete():
+                            nodecontexts[node].action_client(action_client.name, action_client.fmt)
+                            logger.warning(f"Action {action_client._action_name} is incomplete")
+                            logger.debug(
+                                f"goal({action_client.goal}), cancel({action_client.cancel}), "
+                                f"status({action_client.status}), feedback("
+                                f"{action_client.feedback}), result({action_client.result}")
+
+                    else:
+                        action_client.add_unfinished_to_context(nodecontexts[node])
+            self._nodes = nodecontexts
+            return self.summarise()
+        except TimeoutError:
+            logger.exception("failed to connect to the ROS master. Is a robot running?")
```

Comment:
Probably missing a `raise` here. This will implicitly cause the method to return `None`.